source/code-snippets/transactions/txn-callback.js

```diff
@@ -0,0 +1,147 @@
+const { MongoClient } = require('mongodb');
+
+async function cleanUp(client) {
+  try {
+    const customersColl = client.db('testdb').collection('customers');
+    await customersColl.drop();
+  } catch(e) {}
+  try {
+    const inventoryColl = client.db('testdb').collection('inventory');
+    await inventoryColl.drop();
+  } catch(e) {}
+  try {
+    const ordersCollection = client.db('testdb').collection('orders')
+    await ordersCollection.drop();
+  } catch(e) {}
+}
+
+async function setup(client) {
+  try {
+    const customerColl = client.db('testdb').collection('customers');
+    const inventoryColl = client.db('testdb').collection('inventory');
+
+    await customerColl.insertOne({ _id: 98765, orders: [] });
+
+    await inventoryColl.insertMany([
+      { name: 'sunblock', sku: 5432, qty: 85 },
+      { name: 'beach towel', sku: 7865, qty: 41 },
+    ]);
+  } catch (e) {
+    console.log('Unable to insert test data: ' + e);
+  }
+}
+
+async function queryData() {
+  const uri = process.env.MONGODB_URI;
+  const client = new MongoClient(uri, { useUnifiedTopology: true });
+  try {
+    await client.connect();
+    await Promise.all(['customers', 'inventory', 'orders'].map(async c => {
+      const coll = client.db('testdb').collection(c);
+      console.log(JSON.stringify(await coll.find().toArray()));
+    }, client));
+  } finally {
+    client.close();
+  }
+}
+
+// start callback
+async function placeOrder(client, session, cart, payment) {
+  const ordersCollection = client.db('testdb').collection('orders');
+  const orderResult = await ordersCollection.insertOne(
+    {
+      customer: payment.customer,
+      items: cart,
+      total: payment.total,
+    },
+    { session }
+  );
+
+  const inventoryCollection = client.db('testdb').collection('inventory');
+  for (let i=0; i<cart.length; i++) {
+    const item = cart[i];
+
+    // Cancel the transaction when you have insufficient inventory
+    const checkInventory = await inventoryCollection.findOne(
+      {
+        sku: item.sku,
+        qty: { $gte: item.qty }
+      },
+      { session }
+    );
+
+    if (checkInventory === null) {
+      await session.abortTransaction();
+      console.error('Insufficient quantity or SKU not found.');
+    }
+
+    await inventoryCollection.updateOne(
+      { sku: item.sku },
+      { $inc: { 'qty': -item.qty }},
+      { session }
+    );
+  }
+
+  const customerCollection = client.db('testdb').collection('customers');
+  await customerCollection.updateOne(
+    { _id: payment.customer },
+    { $push:  { orders: orderResult.insertedId }},
+    { session }
+  );
+}
+// end callback
+
+const uri = process.env.MONGODB_URI;
+const client = new MongoClient(uri, { useUnifiedTopology: true });
+
+async function run() {
+  /* Test code: uncomment block and commentto run
```

Comment:
Is. This meant to be commented out? If so, just need to fix the typo -> "commentto"

Reply:
Thanks. 
Side comment: Admittedly, this code isn't structured in a way that is easy to either demonstrate the example code nor run the example. Since we never surface it to the user, I think the structure is ok for now.